src/GraphQL/Utility/FileUpload.php

```diff
@@ -0,0 +1,254 @@
+<?php
+
+namespace Drupal\graphql\GraphQL\Utility;
+
+use Drupal\Component\Utility\Bytes;
+use Drupal\Component\Utility\Environment;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\File\FileSystem;
+use Drupal\Core\File\FileSystemInterface;
+use Drupal\Core\Logger\LoggerChannelInterface;
+use Drupal\Core\Session\AccountProxyInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesserInterface;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+use Symfony\Component\Validator\ConstraintViolation;
+
+/**
+ * Service to manage file uploads within GraphQL mutations.
+ */
+class FileUpload {
+
+  use StringTranslationTrait;
+
+  /**
+   * List of violations during file upload.
+   *
+   * @var array
+   */
+  protected $violations = [];
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountProxyInterface
+   */
+  protected $currentUser;
+
+  /**
+   * The mime type guesser service.
+   *
+   * @var \Symfony\Component\HttpFoundation\File\MimeType\MimeTypeGuesserInterface
+   */
+  protected $mimeTypeGuesser;
+
+  /**
+   * The file system service.
+   *
+   * @var \Drupal\Core\File\FileSystemInterface
+   */
+  protected $fileSystem;
+
+  /**
+   * GraphQL logger channel.
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelInterface
+   */
+  protected $logger;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(
+    EntityTypeManagerInterface $entityTypeManager,
+    AccountProxyInterface $currentUser,
+    MimeTypeGuesserInterface $mimeTypeGuesser,
+    FileSystemInterface $fileSystem,
+    LoggerChannelInterface $logger
+  ) {
+    $this->entityTypeManager = $entityTypeManager;
+    $this->currentUser = $currentUser;
+    $this->mimeTypeGuesser = $mimeTypeGuesser;
+    $this->fileSystem = $fileSystem;
+    $this->logger = $logger;
+  }
+
+  /**
+   * Register violations.
+   *
+   * @param mixed $violations
+   *   List of violations.
+   */
+  protected function registerViolations($violations) {
```

Comment:
This is a service. You can't store stuff within it across potentially multiple mutations or you'll leak violations from one mutations into another.

The service itself (if we add one) should only do low level operations worthy of a service that do not assume too many concrete implementation details.

Reply:
Atm, there are only two public methods, `getViolations` and `createTemporaryFileUpload`. The second one is resetting violations at the start of the process. We used it like calling method for creating file and when empty then getting violations. Next mutation can be called without interfering the first one, as the violations got reseted.

But agree this does not look right. So what do you suggest to do here then, how to return violations? Some object? Could return something like FileUploadResponse, would be that ok?

Reply:
will just try something and let's see

Reply:
I'd like to avoid side-effects as much as possible in our code base. Hence, returning a wrapper object instead might make sense indeed.